src/main/java/com/faforever/client/io/DownloadService.java

```diff
@@ -1,49 +1,119 @@
+        log.warn("Could not download file at {}", mirrorUrl);
+      } catch (ChecksumMismatchException e) {
+        log.warn("Checksum did not match for {}", mirrorUrl, e);
+      } catch (IOException e) {
+        log.warn("Could not download file at {}", mirrorUrl, e);
+      }
+    }
+
+    // No mirrors available or they all failed
+    downloadFile(url, targetFile, progressListener, md5sum);
+  }
+
+  /*
+   * Download a file from an URL using a temporary path and copy it to targetFile if it downloaded and the checksum
+   * matched.
+   */
+  public void downloadFile(URL url, Path targetFile, ByteCountListener progressListener, String md5sum) throws IOException, NoSuchAlgorithmException, ChecksumMismatchException {
     Path tempFile = Files.createTempFile(targetFile.getParent(), "download", null);
+    log.debug("Downloading file {} to {}", url, tempFile);
 
     HttpURLConnection urlConnection = (HttpURLConnection) url.openConnection();
 
     ResourceLocks.acquireDownloadLock();
+    MessageDigest messageDigest = MessageDigest.getInstance("MD5");
     try (InputStream inputStream = url.openStream();
+         DigestInputStream digestInputStream = new DigestInputStream(inputStream, messageDigest);
          OutputStream outputStream = Files.newOutputStream(tempFile)) {
 
-      ByteCopier.from(inputStream)
+      ByteCopier.from(digestInputStream)
           .to(outputStream)
           .totalBytes(urlConnection.getContentLength())
           .listener(progressListener)
           .copy();
 
+      // NOTE: It is crucial that we verify the checksum before using the file when downloading from mirrors! We don't
+      // want to be running unverified executables!
+      String checksum = DatatypeConverter.printHexBinary(messageDigest.digest()).toLowerCase();
+      if (!Objects.equals(md5sum, checksum)) {
+        throw new ChecksumMismatchException(url, checksum, md5sum);
+      }
+
       Files.move(tempFile, targetFile, StandardCopyOption.REPLACE_EXISTING);
     } finally {
       ResourceLocks.freeDownloadLock();
       try {
         Files.deleteIfExists(tempFile);
       } catch (IOException e) {
-        log.warn("Could not delete temporary file: " + tempFile.toAbsolutePath(), e);
+        log.warn("Could not delete temporary file: {}", tempFile.toAbsolutePath(), e);
       }
     }
   }
+
+  /*
+   * Get URLs to the same file on all available mirrors.
+   */
+  public List<URL> getMirrorsFor(URL url) {
+    return preferencesService.getPreferences().getMirror().getMirrorURLs().stream()
+        .map(mirror -> getMirrorURL(mirror, url))
+        .flatMap(Optional::stream)
+        .collect(Collectors.toList());
+  }
+
+  public Optional<URL> getMirrorURL(URI mirror, URL url) {
+      URI uri = mirror.resolve(url.getPath());
+      try {
+        return Optional.of(uri.toURL());
+      } catch (MalformedURLException e) {
+        log.warn("Failed to create URL from URI: {}", uri);
+        return Optional.empty();
+      }
```

Comment:
Can we add tests for these as well?

Reply:
I already added tests for these

Reply:
Not sure why it says that line is uncovered because I definitely already have a test that should hit the 'empty' path here. Maybe my test class isn't being picked up for some reason?

Reply:
Could be I will double check and look at the tests. Codecov can be spotty at times

Reply:
I do get this error when trying to run the test through IDEA:
```
> No tests found for given includes: [com.faforever.client.io.DownloadServiceTest.testDownloadFile](filter.includeTestsMatching)
```